### ISSUE TITLE
SqliteException doesn't return argument values

### ISSUE THREAD

**Issue by ebelevics** [simolus3/sqlite3.dart#126]
`final row = await db.into(_table).insertReturning(item, onConflict: DoUpdate((_) => item, target: uniqueColumns));`

On receiving SqliteException (_ON CONFLICT clause does not match any PRIMARY KEY or UNIQUE constraint, SQL logic error (code 1)_) I get error message and causing statement, but I don't get argument values that is crucial to analyse which particular data causes this exception. At end of statement it just returns "RETURNING *". Is it possible to log args on exception so I could report it to crash analytics?

So far it seems args are collected but not logged in:
```dart
  @override
  Future<QueryResult> runSelect(String statement, List<Object?> args) async {
    final stmt = database.prepare(statement, checkNoTail: true);
    try {
      final result = stmt.select(args);
      return QueryResult.fromRows(result.toList());
    } finally {
      stmt.dispose();
    }
  }
```

**Comment by simolus3** [comment]
I've moved this issue to the sqlite3 package as that one is responsible for creating the exceptions.

> but I don't get argument values that is crucial to analyse which particular data causes this exception.

In the particular query you've posted, it's not. You get the error because columns listed in the `ON CONFLICT` clause don't exactly match the primary key or a unique key. This should not happen with drift, perhaps your database schema has diverted from what drift expects. But you'd get that exception with every parameter.

I'm not opposed to adding that information, but I also wonder what kind of SQL logic errors would be triggered by the parameters only. I'm sure there are some, but I can't think of a good case where dumping all parameters is helpful information.

**Comment by ebelevics** [comment]
Alright, to be more concrete I'll share more info about the issue with arbitrarily values:

This is where I put given data.

```dart
class TableA extends Table {
  String get tableName => 'custom_a';

  IntColumn get localId => integer().autoIncrement()();

  TextColumn get fieldA => text().references(TableB, #fieldA, onDelete: KeyAction.cascade)();
  TextColumn get fieldB => text()();
  TextColumn get fieldC => text().nullable()();

  ...(other non-unique fields)

  @override
  List<String> get customConstraints => ['UNIQUE (fieldA, fieldB, fieldC)'];
}
```

This is table specific DAO where make data insert execution:
`  await super.insert_(item, uniqueColumns: [tableA.fieldA, tableA.fieldB, tableA.fieldC]);`

And that DAO is extended with BaseDao, which shares this method:
```dart
  Future<int> insert_(Insertable<D> item, {List<Column>? uniqueColumns}) async {
    final row = await db.into(_table).insertReturning(item, onConflict: DoUpdate((_) => item, target: uniqueColumns));
    return row.toJson()["localId"];
  }
```

The ON CONFLICT error was thrown because I included only [tableA.fieldA] as an example, which is logical as I have  `List<String> get customConstraints => ['UNIQUE (fieldA, fieldB, fieldC)'];`

But for some reason for me it all works, for client it did work, but then suddenly got this ON CONFLICT exception, even tho we share the same app build number, and I can't explain what exactly causes it.

**Comment by ebelevics** [comment]
So I was looking on crash analytics and found that for some reason `ON CONFLICT ` was only referenced one unique field, that explains this issue. Maybe there was this change that was build in same build number.

`SqliteException: SqliteException(1): ON CONFLICT clause does not match any PRIMARY KEY or UNIQUE constraint, SQL logic error (code 1) Causing statement: INSERT INTO custom_a (field_a, field_b, field_c, field_d, field_e, field_f, field_g, field_h, field_i, field_j, field_k, field_l) VALUES (?, ?, NULL, ?, ?, ?, ?, ?, ?, ?, ?, ?) ON CONFLICT(field_a) DO UPDATE SET field_a = ?, field_b = ?, field_c = NULL, field_d = ?, field_e = ?, field_f = ?, field_g = ?, field_h = ?, field_i = ?, field_j = ?, field_k = ?, field_l = ? RETURNING *. Error thrown null.`

**Comment by ebelevics** [comment]
Still even tho probably this specific issue is solved, I would still like to have value arguments logged on Exception as it gives better context on DatabaseException.
For example sqflite, if not mistaken, returns on calling rawQuery SqfliteDatabaseException, that does include argument values. And can't wrap my head around why Drift acts differently as it's basically a wrapper to sqlflite. Is prepare() custom made implementation?

**Comment by ebelevics** [comment]
Today I stumbled with same problem as client showed in video exactly what caused this problem, and thankfully I could replicate.
As error message I get:
```sql
SqliteException: SqliteException(1555): UNIQUE constraint failed: custom_table.local_id, constraint failed (code 1555)
  Causing statement: INSERT INTO "package_data_wh" ("local_id", "field_a", "field_b", "field_c", "field_d", "field_e", "field_f", "field_g", "field_h", "field_i", "field_j", "field_k", "field_l") 
VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) ON CONFLICT("field_a", "field_b", "field_c") DO UPDATE SET "local_id" = ?, "field_a" = ?, "field_b" = ?, "field_c" = ?, "field_d" = ?, "field_e" = ?, "field_f" = ?, "field_g" = ?, "field_h" = ?, "field_i" = ?, "field_j" = ?, "field_k" = ?, "field_l" = ? RETURNING *
```

As locally tested logStatement I get:
```sql
I/flutter (18217): Drift: Sent INSERT INTO "custom_table.local_id" ("local_id", "field_a", "field_b", "field_c", "field_d", "field_e", "field_f", "field_g", "field_h", "field_i", "field_j", "field_k", "field_l") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) ON CON
I/flutter (18217): FLICT("field_a", "field_b", "field_c") DO UPDATE SET "local_id" = ?, "field_a" = ?, "field_b" = ?, "field_c" = ?, "field_d" = ?, "field_e" = ?, "field_f" = ?, "field_g" = ?, "field_h" = ?, "field_i" = ?, "field_j" = ?, "field_k" = ?, "field_l"
I/flutter (18217): = ? RETURNING * with args [7, UX13596219, SA01568585, null, 1, , A, TL, R50, TRANSIT, RS, RS, 2022-10-19T04:50:39.239911Z, 7, UX13596219, SA01568585, null, 1, , A, TL, R50, TRANSIT, RS, RS, 2022-10-19T04:50:39.239911Z]
I/flutter (18217): Drift: Sent ROLLBACK TRANSACTION with args []
```

And I can notice two things with locally tested log statements, first for some reason arg values are double amount (which is weird), but more important "field_c" which is unique constraint received null value.
While from https://stackoverflow.com/questions/22699409/sqlite-null-and-unique it seems that you can add unique and nullable, but after removing "field_c" from unique constraints, I don't receive any errors for now.

And this is all done because I could get better context of data, but it was done locally with log statements enabled not from crash analytics server-side. Thankfully client gave me credentials and more details about issue so I could test it myself, but if this case was live?

**Comment by simolus3** [comment]
Thanks for providing more details. I agree with you now, this package should provide information about parameters when throwing an exception.

**Comment by simolus3** [comment]
Fixed in 065e6305a1320f83460cfe4e18c35ed5739b409d.
